src/aws/flb_aws_util.c

```diff
@@ -927,3 +932,81 @@ flb_sds_t flb_get_s3_key(const char *format, time_t time, const char *tag,
         }
         return NULL;
 }
+
+/*
+ * Replace %f and %L in time_key_format with milliseconds and nanoseconds. And get the
+ * time format output and output length with strftime().
+ */
+size_t create_time_formatter(char **out_buf, const char *time_key_format,
+                             struct flb_time *tms)
+{
+    char millisecond_str[FLB_AWS_MILLISECOND_FORMATTER_LENGTH+1];
+    char nanosecond_str[FLB_AWS_NANOSECOND_FORMATTER_LENGTH+1];
+    char *tmp_parsed_time_str;
+    char *tmp_out_buf;
+    size_t out_size;
+    size_t tmp_parsed_time_str_len;
+    size_t time_key_format_len;
+    struct tm timestamp;
+    struct tm *tmp;
+
+    /*
+     * Guess the max_length needed for tmp_parsed_time_str and tmp_out_buf. The
+     * upper bond is 12*strlen(time_key_format) because the worst scenario will be only
+     * %c in time_key_format, and %c will be tansfer to 24 chars long time_format
+     * by function strftime().
+     */
+    time_key_format_len = strlen(time_key_format);
+    tmp_parsed_time_str_len = 12*time_key_format_len;
+    /*
```

Comment:
As far as I can tell there's always a new line before a comment in the codebase unless the comment is the first line inside a {}, so I would say a new line before this comment would be good for consistency